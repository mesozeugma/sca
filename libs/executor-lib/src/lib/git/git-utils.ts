import { GitLog } from './interfaces/git-log';

export class GitUtils {
  getYearlyCommits(commits: GitLog[], years: number): GitLog[] {
    const result: GitLog[] = [];
    const existingYears: number[] = [];
    for (const commit of commits) {
      const commitYear = commit.createdAt.getFullYear();
      if (!existingYears.includes(commitYear)) {
        result.push(commit);
        existingYears.push(commitYear);
      }
      if (existingYears.length === years) {
        return result;
      }
    }
    return result;
  }
}
