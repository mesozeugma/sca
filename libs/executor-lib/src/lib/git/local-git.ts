import { SimpleGit, simpleGit } from 'simple-git';
import { z } from 'zod';

import { Git } from './interfaces/git';

const LocalGitLog = z
  .object({
    date: z.string().pipe(z.coerce.date()),
    hash: z.string(),
    message: z.string(),
  })
  .transform((v) => ({
    createdAt: v.date,
    hash: v.hash,
    message: v.message,
  }));

export class LocalGit implements Git {
  private git: SimpleGit;

  constructor(repositoryPath: string) {
    this.git = simpleGit({ baseDir: repositoryPath });
  }

  async getHistory() {
    const log = await this.git.log({ strictDate: true });
    this.git.log();
    return log.all.map((e) => LocalGitLog.parse(e));
  }

  async getLatestCommit() {
    const log = await this.git.log({
      maxCount: 1,
      strictDate: true,
    });
    return LocalGitLog.parse(log.latest);
  }
}
