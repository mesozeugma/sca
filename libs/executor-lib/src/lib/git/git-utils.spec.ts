import { GitUtils } from './git-utils';
import { LocalGit } from './local-git';

describe('GitUtils', () => {
  const gitUtils = new GitUtils();

  test('getYearlyCommits', async () => {
    const git = new LocalGit('./libs/executor-lib/code-samples/dissect-cf/');
    const commits = await git.getHistory();
    const result = gitUtils.getYearlyCommits(commits, 5);
    expect(result).toMatchSnapshot();
  });
});
