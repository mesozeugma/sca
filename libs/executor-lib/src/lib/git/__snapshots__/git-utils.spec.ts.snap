// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitUtils getYearlyCommits 1`] = `
[
  {
    "createdAt": 2023-05-25T23:18:02.000Z,
    "hash": "2d0b4b786d0c7113ce76576071e1199ae68e1f02",
    "message": "Refactoring and simplification of tests and some minor apis",
  },
  {
    "createdAt": 2022-09-12T22:03:39.000Z,
    "hash": "2a0e8beceb65a6c1fbacef8e05d02eea400b18ba",
    "message": "InfrastructureModel's fitness is now calculated with a more functional approach",
  },
  {
    "createdAt": 2020-06-24T16:42:27.000Z,
    "hash": "4aebd2a429f47ffe8d6663cf28a25fc4afdabb49",
    "message": "Code quality marker, travis link to the front",
  },
  {
    "createdAt": 2019-05-16T16:39:13.000Z,
    "hash": "4325ccc6a3b280b60bf3642ff5699c0a4c524b9e",
    "message": "FFBFD improver applied in a consolidator",
  },
  {
    "createdAt": 2018-11-11T21:17:51.000Z,
    "hash": "c9008738dbac47af406a1a3db2db3cec3bbca0ab",
    "message": "PSO improvements about final selection and bugfixes",
  },
]
`;
